### ISSUE TITLE
Unhandled Promise Rejection in elasticsearch node client after APM loads.

### ISSUE THREAD

**Issue by bgardner87** [elastic/apm-agent-nodejs#860]
**Describe the bug**
When querying elasticsearch for a document, if that document doesn't exist, the client throws an exception after Elastic responds with a 404. When APM is enabled, this results in an unhandled promise rejection.

**To Reproduce**
1. Sample repo here https://github.com/bgardner87/apm-error-example
2. Fill in appropriate config for your environment in config.js
3. Set up an index in elastic named apm_test with no documents in it. ("PUT apm_test" in kibana dev console will suffice) 
4. run node index.js to start up a HapiJS server with a single route that queries elastic.
5. curl localhost:3000 to trigger the api route.
6. Try with apm enabled (from config) and with apm disabled

**Expected behavior**
The sample catches all exceptions, logs it to console, and moves on. Expectation is that an unhandled promise rejection shouldn't occur.

**Environment (please complete the following information)**
- OS: Have tried on Mac OSX Mojave and Ubuntu 18.x
- Node.js version: 10.15.1 and 8.9.0
- APM Server version: latest in elastic cloud
- Agent version: 2.5.1

**How are you starting the agent? (please tick one of the boxes)**
- [X] Calling `agent.start()` directly (e.g. `require('elastic-apm-node').start(...)`)
- [ ] Requiring `elastic-apm-node/start` from within the source code
- [ ] Starting node with `-r elastic-apm-node/start`

**Additional context**

**Sample output with APM DISABLED**

```
Elastic APM agent is inactive due to configuration
starting
Server running at: http://localhost:3000
creating client
querying
error
{ Not Found :: {"path":"/apm_test/_doc/12","query":{},"statusCode":404,"response":"{\"_index\":\"apm_test\",\"_type\":\"_doc\",\"_id\":\"12\",\"found\":false}"}
    at respond (/Users/bradgardner/dev/apm-error-test/node_modules/elasticsearch/src/lib/transport.js:308:15)
    at checkRespForFailure (/Users/bradgardner/dev/apm-error-test/node_modules/elasticsearch/src/lib/transport.js:267:7)
    at HttpConnector.<anonymous> (/Users/bradgardner/dev/apm-error-test/node_modules/elasticsearch/src/lib/connectors/http.js:166:7)
    at IncomingMessage.wrapper (/Users/bradgardner/dev/apm-error-test/node_modules/lodash/lodash.js:4935:19)
    at emitNone (events.js:111:20)
    at IncomingMessage.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
  status: 404,
  displayName: 'NotFound',
  message: 'Not Found',
  path: '/apm_test/_doc/12',
  query: {},
  body: { _index: 'apm_test', _type: '_doc', _id: '12', found: false },
  statusCode: 404,
  response: '{"_index":"apm_test","_type":"_doc","_id":"12","found":false}',
  toString: [Function],
  toJSON: [Function] }
```

**Sample output with APM ENABLED**
```
starting
Server running at: http://localhost:3000
creating client
querying
error
{ Not Found :: {"path":"/apm_test/_doc/12","query":{},"statusCode":404,"response":"{\"_index\":\"apm_test\",\"_type\":\"_doc\",\"_id\":\"12\",\"found\":false}"}
    at respond (/Users/bradgardner/dev/apm-error-test/node_modules/elasticsearch/src/lib/transport.js:308:15)
    at checkRespForFailure (/Users/bradgardner/dev/apm-error-test/node_modules/elasticsearch/src/lib/transport.js:267:7)
    at HttpConnector.<anonymous> (/Users/bradgardner/dev/apm-error-test/node_modules/elasticsearch/src/lib/connectors/http.js:166:7)
    at IncomingMessage.wrapper (/Users/bradgardner/dev/apm-error-test/node_modules/lodash/lodash.js:4935:19)
    at emitNone (events.js:111:20)
    at IncomingMessage.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
  status: 404,
  displayName: 'NotFound',
  message: 'Not Found',
  path: '/apm_test/_doc/12',
  query: {},
  body: { _index: 'apm_test', _type: '_doc', _id: '12', found: false },
  statusCode: 404,
  response: '{"_index":"apm_test","_type":"_doc","_id":"12","found":false}',
  toString: [Function],
  toJSON: [Function] }
(node:79866) UnhandledPromiseRejectionWarning: Not Found :: {"path":"/apm_test/_doc/12","query":{},"statusCode":404,"response":"{\"_index\":\"apm_test\",\"_type\":\"_doc\",\"_id\":\"12\",\"found\":false}"}
    at respond (/Users/bradgardner/dev/apm-error-test/node_modules/elasticsearch/src/lib/transport.js:308:15)
    at checkRespForFailure (/Users/bradgardner/dev/apm-error-test/node_modules/elasticsearch/src/lib/transport.js:267:7)
    at HttpConnector.<anonymous> (/Users/bradgardner/dev/apm-error-test/node_modules/elasticsearch/src/lib/connectors/http.js:166:7)
    at IncomingMessage.wrapper (/Users/bradgardner/dev/apm-error-test/node_modules/lodash/lodash.js:4935:19)
    at emitNone (events.js:111:20)
    at IncomingMessage.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
(node:79866) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:79866) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

- Agent config options 
  <details>
    <summary>Click to expand</summary>

    ```
    serviceName: config.apmService,
    secretToken: config.apmToken,
    serverUrl: config.apmUrl,
    captureBody: 'all',
    logLevel: 'info',
    active: config.apmEnabled
    ```
  </details>
- `package.json` dependencies:
  <details>
    <summary>Click to expand</summary>

    ```
    "elastic-apm-node": "^2.5.1",
    "elasticsearch": "^15.4.1",
    "hapi": "^18.1.0"
    ```
  </details>

**Comment by bgardner87** [comment]
Looking a little closer:

https://github.com/elastic/apm-agent-nodejs/blob/b80aa0485dd62b25da015c02a0d96d509251e50c/lib/instrumentation/modules/elasticsearch.js#L44

Changing
```
p.then(function () {
    span.end()
})
```

To
```
p.then(function () {
    span.end()
}).catch(function() {
    span.end()
})
```
Stops the unhandled promise rejection, and doesn't seem to impact the caller's ability to catch the error, but I'm not totally sure that I understand it fully yet. I'm happy to make a PR for it and take the discussion there if that's appropriate here.

**Comment by Qard** [comment]
Yep, that looks like a valid fix to me. If you want to open a PR for that, I'm happy to get that landed and released.

**Comment by bgardner87** [comment]
@Qard Thanks for the sanity check. I'll submit a PR shortly.

I'm still at a bit of a loss as to why this is necessary though. The function returns the original promise so in my head everything regarding the promise chain should be in tact? Mostly asking out of curiosity at this point if you can offer any insight.

**Comment by bgardner87** [comment]
NVM - worked on it a bit further and believe I have it figured out. I'll put the  PR up in just a bit.

Thanks!

**Comment by Qard** [comment]
It's because each call to `promise.then(...)` produces a new promise. It is creating a promise chain internally, with that `then(...)` call, which never has a `catch(...)` call. The returned promise is the parent promise, so calling `then(...)` on that produces a new branch which will have a separate reject handler.

**Comment by bgardner87** [comment]
@Qard thanks for the specific explanation, we came to the same conclusion on reviewing it internally as well. Working on the PR now!

**Comment by watson** [comment]
A fix for this was just released in v2.7.0
